sdk/scheduler/src/main/java/org/apache/mesos/offer/constrain/RoundRobinByHostnameRule.java

```diff
@@ -0,0 +1,164 @@
+package org.apache.mesos.offer.constrain;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+import org.apache.commons.lang3.builder.EqualsBuilder;
+import org.apache.commons.lang3.builder.HashCodeBuilder;
+import org.apache.mesos.Protos.Offer;
+import org.apache.mesos.Protos.TaskInfo;
+import org.apache.mesos.offer.OfferRequirement;
+import org.apache.mesos.offer.TaskException;
+import org.apache.mesos.offer.TaskUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+/**
+ * Implements logic for Marathon's GROUP_BY operator for hostnames. Ensures that tasks are evenly
+ * distributed across agents in the system as they are rolled out.
+ *
+ * Example:
+ *  hostname |     tasks
+ * ----------+---------------
+ *   host-1  | a-1, b-1, c-1
+ *   host-2  | a-2, c-2, c-3
+ *   host-3  | b-2, c-4
+ * Result:
+ *  allow (only) host-3, unless we know that there's >=4 hosts via the agent_count parameter
+ *
+ * Example:
+ *  hostname |     tasks
+ * ----------+---------------
+ *   host-1  | a-1, b-1, c-1
+ *   host-2  | a-2, c-2, c-3
+ *   host-3  | b-2, c-4, b-3
+ * Result:
+ *  allow any of host-1/host-2/host-3, unless we know that there's >=4 hosts via the agent_count
+ *  parameter.
+ */
+public class RoundRobinByHostnameRule implements PlacementRule {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(RoundRobinByHostnameRule.class);
+
+    private final Optional<Integer> agentCountOptional;
+    private final StringMatcher taskFilter;
+
+    public RoundRobinByHostnameRule(Optional<Integer> agentCount) {
+        this(agentCount, null);
+    }
+
+    @JsonCreator
+    public RoundRobinByHostnameRule(
+            @JsonProperty("agent_count") Optional<Integer> agentCount,
+            @JsonProperty("task_filter") StringMatcher taskFilter) {
+        this.agentCountOptional = agentCount;
+        if (taskFilter == null) { // null when unspecified in serialized data
+            taskFilter = AnyMatcher.create();
+        }
+        this.taskFilter = taskFilter;
+    }
+
+    @Override
+    public Offer filter(Offer offer, OfferRequirement offerRequirement, Collection<TaskInfo> tasks) {
```

Comment:
It feels like this logic should be generalizable between this and the attribute round-robin, maybe by making this a higher-order function.